### DISCUSSION TITLE
OCPP testing tools

### DISCUSSION THREAD

**Discussion by bitbs566** [lbbrhzn/ocpp#496] [Q&A]
There are several testing/simulations charging point tools floating around in the GitHub aimed to test the functionality of your OCPP server, but unfortunately it seems that neither of them working out of the box correctly. Can anyone suggest good program to test the functionality of the OCPP HA integration while Im waiting my for my OCPP EVSE controller? 

Im always get this error in HA:
Logger: custom_components.ocpp
Source: custom_components/ocpp/api.py:205
Protocols mismatched | expected Subprotocols: ['ocpp1.6'], but client supports ocpp1.6, ocpp1.5 | Closing connection

I was able to correct the main.js file on the other side to try to fool OCPP HA server into thinking that it is sending only 1.6 signals, but it doesn't work unfortunately. Does anyone have any ideas? I will be glad to hear anything from you guys, thanks in advance.

**Comment by lbbrhzn** [comment]
This looks like a bug on our side. The client indicates it supports both ocpp1.5 and ocpp1.6. The HA integration supports ocpp1.6 only, so it should accept the connection.

**Reply by bitbs566** [reply]
Thanks for your fast reply,
yes I did realise this, that there is an issue on the client side, but even if I have tried to correct it, it doesn't work still. Are you aware of any good OCPP client simulators that should work with your integration flawlessly? Thanks.

**Reply by lbbrhzn** [reply]
We use our own client simulator for testing, as part of our continuous integration workflow.  This is based on the same ocpp python package. You can find it in the tests folder.

It would be good to test the integration with a third party simulator. The best one would be the official one that is used for OCPP qualification, but that is not free.  I've looked at a free alternative in the past, but found it complicated to set up:
https://github.com/oglimmer/scriptable-ocpp-chargepoint-simulator

**Reply by drc38** [reply]
I received this in my inbox recently: 
>The Open Charge Alliance proudly presents the second combined OCPP 2.0.1 and 1.6 Plugfest on June 9 & 10 2022 ! The Plugfest gives developers from all over the world a chance to test their OCPP devices with eachother.

>You can now register your Charging Station or Charging Station Management system.

>[Register Here](https://oca.lt.acemlnb.com/Prod/link-tracker?redirectUrl=aHR0cHMlM0ElMkYlMkZ3d3cub3BlbmNoYXJnZWFsbGlhbmNlLm9yZyUyRm9jcHAtMjAxLWFuZC0xNi1vbmxpbmUtcGx1Z2Zlc3QtanVuZS05LWVuLTEwLTIwMjIlMkY=&sig=3XKoYsNpNrw13V2APbhoBHpye993JWeR4WjT4LEdCdB6&iat=1650386234&a=%7C%7C67253792%7C%7C&account=oca%2Eactivehosted%2Ecom&email=lfNSdZDCNJL8jjOW9p%2F6FnwFoqDlMHNmyq65fGLdufk%3D&s=2c02d0796d8cbfa84a04c8a26c0292c7&i=21A31A3A252)

**Reply by lbbrhzn** [reply]
@drc38 cool, I think i'll register, just for the fun of it. got time to join as well?

**Reply by drc38** [reply]
@lbbrhzn  nice one, let me know how it goes. I'm not setup to have an internet-based connection into my HA server. Perhaps we look at adding secure websocket beforehand? Did you see my suggestion on being able to override FeatureProfiles on setup?

**Reply by lbbrhzn** [reply]
I can use port forwarding and reverse proxy allow both secure and non-secure connections straight into my visual studio container.
Would be nice to add support for multiple chargepoints beforehand.

Yes, I saw your suggestion for overriding the FeatureProfiles, I'll add that to the config flow. Seems we are adding more and more features to allow non-compliant implementations..

**Reply by drc38** [reply]
My cynical take is more manufacturers are making tweaks to lock customers into their CMS solution rather than compete on price, features and service...

**Reply by drc38** [reply]
> @drc38 cool, I think i'll register, just for the fun of it. got time to join as well?

@lbbrhzn did you manage to join the Plugfest?
